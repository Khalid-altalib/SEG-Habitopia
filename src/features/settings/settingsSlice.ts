import { createSlice, createAsyncThunk, PayloadAction } from "@reduxjs/toolkit";
import { DataStore } from "aws-amplify";
import { Toast } from "react-native-toast-message/lib/src/Toast";
import { Settings } from "../../../types";
import { getUserFromDatabase } from "../../app/util";
import { User } from "../../models";
import { updatePassword } from "../auth/authSlice";

export type SettingsState = {
  settings: Settings;
  fetchSettings: {
    loading: boolean;
    error: string;
  };
  setSettings: {
    loading: boolean;
    error: string;
  };
};

const initialState: SettingsState = {
  settings: {
    email: "",
    password: "",
    notifications: false,
    name: "",
    biography: "",
  },
  fetchSettings: {
    loading: false,
    error: "",
  },
  setSettings: {
    loading: false,
    error: "",
  },
};

export const fetchSettings = createAsyncThunk<
  Settings,
  void,
  { rejectValue: string }
>("settings/fetch", async (_, thunkAPI) => {
  try {
    const userData  = await getUserFromDatabase(thunkAPI);

    const response =  {
      email: userData.email,
      name: userData.name,
      notifications: userData.notifications,
      biography: userData.biography,
    };
   
    return (await response) as Settings;
  } catch (error: any) {
    const message = error.message;
    return thunkAPI.rejectWithValue(message);
  }
});


export const setSettings = createAsyncThunk<
  void,
  object,
  { rejectValue: string }
>("settings/set", async (settings: any, thunkAPI) => {
  try {
    const {name, notifications, biography, password, oldPassword} = settings;
    const user = await getUserFromDatabase(thunkAPI);
    // update the one that is not null
    await DataStore.save(
      User.copyOf(user, (updated) => {
        if (name !== undefined) {
          updated.name = name;
        }
        if (notifications !== undefined) {
          updated.notifications = notifications;
        }
        if (biography !== undefined) {
          updated.biography = biography;
        }
        if (password !== undefined) {
          updatePassword(password, oldPassword);
        }
      }));

    Toast.show({type: "success", text1: "Your new setting has been saved! âœ…"})

    return settings; 
  } catch (error: any) {
    const message = error.message;
    return thunkAPI.rejectWithValue(message);
  }
});

export const settingsSlice = createSlice({
  name: "settings",
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(
      fetchSettings.fulfilled,
      (state: SettingsState, action: PayloadAction<Settings>) => {
        state.settings = action.payload;
        state.fetchSettings.loading = false;
        state.fetchSettings.error = "";
      }
    );
    builder.addCase(fetchSettings.pending, (state: SettingsState) => {
      state.fetchSettings.loading = true;
      state.fetchSettings.error = "";
    });
    builder.addCase(
      fetchSettings.rejected,
      (state: SettingsState, action: PayloadAction<any>) => {
        state.fetchSettings.loading = false;
        state.fetchSettings.error = action.payload.message;
      }
    );
    builder.addCase(
      setSettings.fulfilled,
      (state: SettingsState, action: PayloadAction<any>) => {
        state.settings = { ...state.settings, ...action.payload };
        state.setSettings.loading = false;
      }
    );
    builder.addCase(setSettings.pending, (state: SettingsState) => {
      state.setSettings.loading = true;
    });
    builder.addCase(setSettings.rejected, (state: SettingsState) => {
      state.setSettings.loading = false;
    });
  },
});

export const {} = settingsSlice.actions;

export default settingsSlice.reducer;
