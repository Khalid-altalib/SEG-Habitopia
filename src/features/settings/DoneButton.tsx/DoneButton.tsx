import { NavigationProp, useNavigation } from "@react-navigation/native";
import { Button, View } from "native-base";
import React from "react";
import { RootParams } from "../../../../types";
import { useAppDispatch, useAppSelector } from "../../../app/hooks";
import { setSettings } from "../settingsSlice";

type Props = {
  disabled?: boolean;
  getValues: (valueName: string) => string;
  valueName: string;
};

const DoneButton = (props: Props) => {
  const navigation = useNavigation<NavigationProp<RootParams>>();
  const { getValues, disabled, valueName } = props;

  const { loading } = useAppSelector((state) => state.settings.setSettings);

  const dispatch = useAppDispatch();

  const handlePress = async () => {
    await dispatch(setSettings({ [valueName]: getValues("formValue") }));
    navigation.goBack();
  };

  return (
    <View>
      <Button
        onPress={handlePress}
        mt={8}
        isDisabled={disabled}
        isLoading={loading}
        testID="button"
      >
        Done
      </Button>
    </View>
  );
};

export default DoneButton;
