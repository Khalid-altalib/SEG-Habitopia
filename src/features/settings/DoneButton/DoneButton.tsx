import Button from "@components/Button";
import StatusContainer from "@components/StatusContainer/StatusContainer";
import { NavigationProp, useNavigation } from "@react-navigation/native";
import { View } from "native-base";
import React from "react";
import { ButtonType, RootParams } from "../../../../types";
import { useDispatch, useSelector } from "../../../app/hooks";
import { setSettings } from "../settingsSlice";

type Props = {
  getValues: (valueName: string) => string;
  valueName: string;
  oldPassword?: string;
};

const DoneButton = (props: Props) => {
  const navigation = useNavigation<NavigationProp<RootParams>>();
  const { getValues, valueName, oldPassword } = props;

  const { loading } = useSelector((state) => state.settings.setSettings);

  const dispatch = useDispatch();

  const handlePress = async () => {
    await dispatch(
      setSettings({
        [valueName]: getValues("formValue"),
        oldPassword: oldPassword,
      })
    );
    if (navigation.canGoBack()) {
      navigation.goBack();
    }
  };

  return (
    <Button
      type={ButtonType.Primary}
      onPress={handlePress}
      isFullWidth
      isLoading={loading}
      style={{ marginTop: 25 }}
    >
      Done
    </Button>
  );
};

export default DoneButton;
