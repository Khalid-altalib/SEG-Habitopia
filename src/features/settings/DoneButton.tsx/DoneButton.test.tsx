import { mount, ReactWrapper } from "enzyme";
import { Provider } from "react-redux";
import { NativeBaseProvider } from "native-base";
import configureStore, { MockStore } from "redux-mock-store";
import thunk from "redux-thunk";
import { setSettings, SettingsState } from "../settingsSlice";
import DoneButton from "./DoneButton";
import { AnyAction } from "redux";

describe("DoneButton", () => {
  const getValuesMock = jest.fn();
  const valueName = "email";
  const navigation = { goBack: jest.fn() } as any;
  let store: MockStore;
  let wrapper: ReactWrapper;

  const initialState: SettingsState = {
    settings: {
      email: "example@gmail.com",
      password: "",
      notifications: true,
      name: "John Doe",
    },
    fetchSettings: {
      loading: false,
      error: "",
    },
    setSettings: {
      loading: false,
      error: "",
    },
  };

  beforeEach(() => {
    store = configureStore([thunk])({ settings: initialState });
    wrapper = mount(
      <NativeBaseProvider>
        <Provider store={store}>
          <DoneButton
            disabled={false}
            getValues={getValuesMock}
            valueName={valueName}
          />
        </Provider>
      </NativeBaseProvider>
    );
  });

  afterEach(() => {
    jest.clearAllMocks();
    wrapper.unmount();
  });

  it("renders correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("dispatches setSettings action when pressed", async () => {
    const mockPayload = { email: "test@test.com" };
    getValuesMock.mockReturnValue("test@test.com");
    store.dispatch = jest.fn().mockResolvedValueOnce(mockPayload);
    const button = wrapper.find("Button");
    console.log(button);
    await button.simulate("mouseDown");

    expect(store.dispatch).toHaveBeenCalledWith(setSettings(mockPayload));
    expect(navigation.goBack).toHaveBeenCalled();
  });

  //   it("disables button when disabled prop is true", () => {
  //     const disabledWrapper = mount(
  //       <NativeBaseProvider>
  //         <Provider store={store}>
  //           <DoneButton
  //             disabled={true}
  //             getValues={getValuesMock}
  //             valueName={valueName}
  //           />
  //         </Provider>
  //       </NativeBaseProvider>
  //     );

  //     const button = disabledWrapper.find("Button");
  //     expect(button.props().disabled).toBe(true);
  //   });
});
