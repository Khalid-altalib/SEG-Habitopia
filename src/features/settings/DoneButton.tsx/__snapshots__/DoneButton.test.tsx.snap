// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DoneButton renders correctly 1`] = `ReactWrapper {}`;
